gensim/matutils.py

```diff
@@ -692,34 +692,42 @@ def unitvec(vec, norm='l2', return_norm=False):
     """
     if norm not in ('l1', 'l2'):
         raise ValueError("'%s' is not a supported norm. Currently supported norms are 'l1' and 'l2'." % norm)
+
     if scipy.sparse.issparse(vec):
         vec = vec.tocsr()
         if norm == 'l1':
             veclen = np.sum(np.abs(vec.data))
         if norm == 'l2':
             veclen = np.sqrt(np.sum(vec.data ** 2))
         if veclen > 0.0:
+            if np.issubdtype(vec.dtype, np.int):
+                vec = vec.astype(np.float) / veclen
+            else:
+                vec /= veclen
+                vec = vec.astype(vec.dtype)
```

Comment:
Slightly confusing. For consistency, I'd prefer -
```python
if np.issubdtype(vec.dtype, np.int):
    vec = vec.astype(np.float) / veclen
else:
    vec = vec.astype(vec.dtype) / veclen
```
Does that make sense?
In fact, if dividing by `veclen` cannot change the dtype (is this the case?), even something like -
```python
if np.issubdtype(vec.dtype, np.int):
    vec = vec.astype(np.float)
vec /= veclen
```

Reply:
Also, sorry I'm really nitpicking here :)
But small things like this cause a slow decline in overall code quality over time.

Reply:
Unfortunately dividing by `veclen` actually changes the dtype, causing the original problem to manifest itself (i.e. float32 inputs outputting float64). Even trying something like this:

```
if np.issubdtype(vec.dtype, np.int):
     vec = vec.astype(np.float) / veclen
else:
     vec = vec.astype(vec.dtype) / veclen.astype(vec.dtype)
```

causes the same problem. This is why I divide by `veclen` and then cast the dtype. Do you have any suggestions to get around this?

Reply:
Actually, I really like your second suggestion, and it passes all tests:

```
if np.issubdtype(vec.dtype, np.int):
    vec = vec.astype(np.float) 
vec /= veclen
```

That's very neat indeed, I'll commit it.

Reply:
Also, don't worry about the nitpicking! When I am experienced enough to pick up on such things I'm sure I will be nitpicking like this haha.

Reply:
Thanks for the fix and the explanation! Looks good.